session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift

```diff
@@ -0,0 +1,231 @@
+    let adds: [Add]
+    /// Elements to update (change data).
+    let updates: [Update]
+    /// Elements to remove.
+    let removes: [Remove]
+
+    /// If diff has no changes.
+    /// Empty diff indicates that arrays compared in `computeDiff(oldArray:newArray:)` are equal.
+    var isEmpty: Bool {
+        return adds.isEmpty && updates.isEmpty && removes.isEmpty
+    }
+}
+
+/// Thrown if unexpected happened in implemented algorithm.
+internal struct DiffError: Error {}
+
+/// An item in symbols table (`[DiffableID: Symbol]`) in Heckel's algorithm.
+///
+/// Note: Unlike original algorithm, our implementation doesn't use `OC`, `NC` and `OLNO` counters. In our case, elements within each array are unique,
+/// so instead of repetition counters, we define basic `Bool` flags to track occurence of certain `id` in one or both files.
+private struct Symbol {
+    /// If element with certain `id` occurs in `oldArray`.
+    var inOld: Bool
+    /// If element with certain `id` occurs in `newArray`.
+    var inNew: Bool
+    /// The index of element in `oldArray`.
+    var indexInOld: Int?
+}
+
+/// An entry in `oa` (old array) and `na` (new array) arrays in Heckel's algorithm.
+private enum Entry: Equatable {
+    /// Reference to `Symbol` in `table: [DiffableID: Symbol]`.
+    case reference(DiffableID)
+    /// Index of element in other array (in `oldArray` for `na: [Entry]` and in `newArray` for `oa: [Entry]`).
+    case index(Int)
+}
+
+/// Computes a diff between two arrays.
+///
+/// This implementation is based on Paul Heckel's algorithm for finding differences between files. It isolates differences in a way that corresponds
+/// closely to our intuitive notion of difference (it finds the longest common subsequence). It is computationally efficient: `O(n)` in time and memory.
+///
+/// Unlike original Heckel's algorithm, our implementation assumes that elements are unique within each of two arrays. It means that all elements in
+/// `oldArray` are guaranteed to have different `id` (same for `newArray`). Elements with the same `id` can appear in both arrays, which
+/// indicates one of two things determined by `newElement.isDifferent(than: oldElement)`:
+/// - either the element was not altered and can be skipped in diff,
+/// - or the element was changed and it should be reflected in `Diff.Update`.
+///
+/// Like original algorithm, our implementation uses 6 passes over both arrays to determine diff
+///
+/// Ref.:
+/// - _"A technique for isolating differences between files"_ Paul Heckel (1978) - https://dl.acm.org/citation.cfm?id=359467
+///
+/// - Parameters:
+///   - oldArray: original array
+///   - newArray: new array
+/// - Returns: `Diff` describing changes from `oldArray` to `newArray`.
+internal func computeDiff<E: Diffable>(oldArray: [E], newArray: [E]) throws -> Diff<E> {
+    var table: [DiffableID: Symbol] = [:]
+    var oa: [Entry] = [] // old array entries
+    var na: [Entry] = [] // new array entries
```

Comment:
Both `table`, `oa` and `na` names refer to the Heckel's paper and are commonly used among different implementations that I found around. I'd keep it for clarity even if it's not our convention.

Reply:
For algorithm like this, short var name helps for readability 👍